src/ui/groups_widgets.cc

```diff
@@ -871,12 +876,66 @@ void DictGroupsWidget::addAutoGroupsByFolders()
     addUniqueGroup( group );
     DictListModel * model = getCurrentModel();
 
-    for ( const auto& dict : groupToDicts.values( group ) ) {
+    for ( const auto & dict : groupToDicts.values( group ) ) {
       model->addRow( QModelIndex(), dict );
     }
   }
 }
 
+void DictGroupsWidget::groupsByMetadata()
+{
+  if ( activeDicts->empty() ) {
+    return;
+  }
+  if ( QMessageBox::information( this,
+                                 tr( "Confirmation" ),
+                                 tr( "Are you sure you want to generate a set of groups based on metadata.toml?" ),
+                                 QMessageBox::Yes,
+                                 QMessageBox::Cancel )
+       != QMessageBox::Yes ) {
+    return;
+  }
+  // map from GroupName to dicts
+  QMultiMap< QString, sptr< Dictionary::Class > > groupToDicts;
+
+  for ( const auto & dict : *activeDicts ) {
+    auto baseDir = dict->getContainingFolder();
+    if ( baseDir.isEmpty() )
+      continue;
+
+    auto filePath = Utils::Path::combine( baseDir, "metadata.toml" );
+    if ( !QFile::exists( filePath ) ) {
+      qDebug() << "the dictionary folder:" << baseDir << " contain no metadata.toml";
+      continue;
+    }
+    try {
+      auto config = toml::parse_file( filePath.toStdString() );
+
+      toml::array * categories = config.get_as< toml::array >( "category" );
+      if (!categories) {
+        continue;
+      }
+      // increment each element with visit()
+      for ( auto && elem : *categories ) {
+        elem.visit( [ & ]( auto && el ) noexcept {
+          auto group = QString::fromStdString( el.as_string()->value_or( "" ) ).trimmed();
+          if ( group.isEmpty() )
+            return;
+
+          groupToDicts.insert( group, dict );
+        } );
+      }
```

Comment:
Does `elem.visit` really needed?

I stripped it into

```
      for ( auto && elem : *categories ) {
          auto group = QString::fromStdString( elem.as_string()->value_or( "" ) ).trimmed();
          if ( group.isEmpty() )
            return;
          groupToDicts.insert( group, dict );
      }
```

And the code still works OK.

Reply:
yes ,your code is better.  I'm still struggled with the syntax now.

Reply:
The code also identical to 

```
categories->for_each( [ &groupToDicts, &dict ]( auto && elem ) {
  auto group = QString::fromStdString( elem.as_string()->value_or( "" ) ).trimmed();
  if ( group.isEmpty() )
    return;
  groupToDicts.insert( group, dict );
} );
```